### PR TITLE
feat: implement increasing precision for values

### LINKED ISSUE #3977: Implement increasing precision for values
Some storage and memory values are displayed with too little precision, which can be misleading.
Related design ticket: Increase precision for storage and memory values #3975 

Figma mockups: https://www.figma.com/design/QVIkAC32FFvSsrIzEj8Mb8/Storage?node-id=2686-241750
Example:
_"2 of 3 TB" is shown with 45.7%._
Users see “2 of 3” and intuitively expect ~66%, but the actual percentage is correct. The problem is that both used and total values are rounded too aggressively, breaking the visual connection between the fraction and the percentage.

### Main Idea

We want to increase display precision for large values (GB/TB) while keeping the values human-readable.
We will follow the existing Storage Formatting Guidelines (already approved and partially implemented in Database Storage page).

### Current Rules
_to be applied_

Automatic unit selection. Automatically choose the largest suitable unit (KB → MB → GB → TB → PB) where the value is in the comfortable range (usually 1–999, sometimes up to 9999).

Image

**Decimal places (key improvements):**

- TB and above: 1–2 decimal places
- GB: 0–1 decimal place
- Percentages: 1 decimal place when < 1% (e.g. 0.7%)

**Other formatting:**

- Space as thousands separator: 1 234.5 GB
- Always show unit with space: 3.0 TB, 61.2 GB
- In the same section/block — try to use the same unit when possible for easier comparison

Full rules with examples are available in [Figma](https://www.figma.com/design/QVIkAC32FFvSsrIzEj8Mb8/Storage?node-id=2688-329145&t=6z5peHojGMGm7vWh-4).

Image

### Areas to update

- Cluster Overview (main dashboard cards)
- Database Tabs
- Database Storage section (Tablet Storage + Database Storage) — updated in new design by Anton Standrik
- Any other places where GB/TB storage or memory values are shown (if possible)

Image

Image

Image

### PR DESCRIPTION
**Pull request by kkdras**
Resolves #3977 
[stand](http://ydb-vla-dev02-004.search.yandex.net:8765/monitoring/cluster/databases?databasePage=diagnostics)

### Test Status: <span style="color: orange;">⚠️ FLAKY</span>
  📊 [Full Report](https://ydb-platform.github.io/ydb-embedded-ui/4021/?t=1781853968581)

  | Total | Passed | Failed | Flaky | Skipped |
  |:-----:|:------:|:------:|:-----:|:-------:|
  | 688 | 682 | 0 | 3 | 3 |

  😟 No changes in tests. 😕

  ### Bundle Size: 🔺
  Current: 64.15 MB | Main: 64.14 MB
  Diff: +6.08 KB (0.01%)

  ⚠️ Bundle size increased. Please review.

  <details>
  <summary>ℹ️ CI Information</summary>

  - Test recordings for failed tests are available in the full report.
  - Bundle size is measured for the entire 'dist' directory.
  - 📊 indicates links to detailed reports.
  - 🔺 indicates increase, 🔽 decrease, and ✅ no change in bundle size.
  </details>

<h3>Greptile Summary</h3>

This PR introduces precision-aware formatting for metric byte values throughout the UI: values now display with more decimal places when small and fewer when large, and auto-upgrade to the next unit when rounding would reach the rollover boundary (e.g., 999.6 MB → 1 GB).

- Adds `getMetricBytesDisplaySize` with a rollover-aware while-loop that upgrades the display unit when the rounded value would hit 1000, and refines `getConvertedMetricBytesDecimalPlaces` so MB values ≥ 10 drop to 0 decimal places.
- Introduces `getFormatBytesPrecision` / `formatNetworkMetric` for legend and network-throughput fields, and threads a new `FormatValuesPrecision` through `formatValues` / `formatStorageValues` so caller-computed precision reaches the formatter.
- Removes the `percentPrecision` prop from `ClusterMetricsBaseProps` and replaces it with an inline `getOverviewMetricPercentPrecision` rule (1 decimal only below 1 %).

<h3>Confidence Score: 5/5</h3>

Safe to merge — all changed formatting paths are covered by new and updated unit tests, and the e2e snapshot has been regenerated to match the new rounding rules.

The changes are self-contained formatting utilities with no side effects on data fetching or state management. New unit tests cover the rollover boundary, the precision ladder for every byte tier, and edge cases (NaN/undefined/null inputs). The e2e test expectations were updated to match the new MB-rounding behavior. No pre-existing tests were removed or weakened.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/utils/storageMetrics.ts | Core change: adds `getMetricBytesDisplaySize` (rollover loop) and exports `getConvertedMetricBytesDecimalPlaces`; logic is sound and covered by new unit tests |
| src/utils/metrics/formatMetricLegend.ts | Adds `getFormatBytesPrecision` helper and `formatNetworkMetric`; significant-figure arithmetic is correct and well-tested |
| src/utils/dataFormatters/common.ts | Adds `FormatValuesPrecision` interface and threads optional precision through `formatValues`; backwards-compatible with `?? 0` / `?? 1` defaults |
| src/containers/Tenant/Diagnostics/TenantOverview/TenantStorage/displayFormatters.ts | Migrates unit-selection calls from `getBytesSizeUnit` to `getMetricBytesDisplaySize`; `formatTenantStorageTableMetric` now accepts `null` and delegates to `formatByteMetric` |
| src/containers/Cluster/ClusterOverview/utils.tsx | Replaces caller-supplied `percentPrecision` with `getOverviewMetricPercentPrecision` (1 decimal only below 1%); drops `percentPrecision` from `ClusterMetricsBaseProps` |
| src/utils/bytesParsers/formatBytes.ts | Exports `bytesSizes` (ordered keys of `sizes`) to let `storageMetrics.ts` walk the unit ladder; trivial, correct change |

</details>

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["Raw byte value"] --> B["getBytesSizeUnit(value)\n(initial unit guess)"]
    B --> C{"getMetricBytesDisplaySize\nwhile loop"}
    C -->|"rounded >= 1000\n&& nextSize exists"| D["Upgrade to next unit"]
    D --> C
    C -->|"rounded < 1000\nor no nextSize"| E["resolvedSize"]

    E --> F["getConvertedMetricBytesDecimalPlaces\n(size, convertedValue)"]
    F --> G{"size?"}
    G -->|"kb"| H["0 if integer, else 1"]
    G -->|"mb < 10"| I["0 if integer, else 1"]
    G -->|"mb >= 10"| J["0"]
    G -->|"gb"| K["0 if integer;\ngbDecimalPlacesBelowOne if < 1;\nelse 1"]
    G -->|"tb/pb < 10"| L["2"]
    G -->|"tb/pb >= 10"| M["1"]

    H & I & J & K & L & M --> N["decimalPlaces"]
    N --> O["getFormatBytesPrecision\n= trunc(value).digits + decimalPlaces"]
    O --> P["roundToPrecision(convertedValue, precision)\n(significant figures)"]
    P --> Q["formatNumber → display string"]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A["Raw byte value"] --> B["getBytesSizeUnit(value)\n(initial unit guess)"]
    B --> C{"getMetricBytesDisplaySize\nwhile loop"}
    C -->|"rounded >= 1000\n&& nextSize exists"| D["Upgrade to next unit"]
    D --> C
    C -->|"rounded < 1000\nor no nextSize"| E["resolvedSize"]

    E --> F["getConvertedMetricBytesDecimalPlaces\n(size, convertedValue)"]
    F --> G{"size?"}
    G -->|"kb"| H["0 if integer, else 1"]
    G -->|"mb < 10"| I["0 if integer, else 1"]
    G -->|"mb >= 10"| J["0"]
    G -->|"gb"| K["0 if integer;\ngbDecimalPlacesBelowOne if < 1;\nelse 1"]
    G -->|"tb/pb < 10"| L["2"]
    G -->|"tb/pb >= 10"| M["1"]

    H & I & J & K & L & M --> N["decimalPlaces"]
    N --> O["getFormatBytesPrecision\n= trunc(value).digits + decimalPlaces"]
    O --> P["roundToPrecision(convertedValue, precision)\n(significant figures)"]
    P --> Q["formatNumber → display string"]
```

</a>

<sub>Reviews (3): Last reviewed commit: ["refactor: keep numeric value formatter u..."](https://github.com/ydb-platform/ydb-embedded-ui/commit/6c20739ba74201f21e6f2e038c44feae95f8ca22) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38509176)</sub>

### CODE REVIEW

**Review by kkdras — COMMENTED** [review]
@greptile review

**Review by kkdras — COMMENTED** [review]
@codex review

**Comment by kkdras** [comment]
@greptile review

**Comment by kkdras** [comment]
@codex review

**Comment by kkdras** [comment]
@codex review

**Comment by kkdras** [comment]
@greptile review

File: src/containers/Cluster/ClusterOverview/utils.tsx [hunk 1]
```diff
@@ -21,12 +21,15 @@ function calculateFillWidth(value: number, capacity: number) {
     return (value / capacity) * 100;
 }
 
+function getOverviewMetricPercentPrecision(fillWidth: number) {
+    return fillWidth > 0 && fillWidth < 1 ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (kkdras) [reply src/containers/Cluster/ClusterOverview/utils.tsx:25]: This was an intentional behavior change.

The goal of this PR is to align overview percentage formatting with the requirement from issue #3977: show one decimal place only for values below 1%, and show whole percentages otherwise. The current cluster and tenant overview cards now follow the same rule consistently.

File: src/containers/Cluster/ClusterOverview/utils.tsx [hunk 2]
```diff
@@ -36,7 +39,7 @@ export function calculateBaseDiagramValues({
         colorizeProgress,
     });
 
-    const percents = formatPercent(fillWidth / 100, percentPrecision);
+    const percents = formatPercent(fillWidth / 100, getOverviewMetricPercentPrecision(fillWidth));
```
[the remark this answers is not in the document]
Comment (kkdras) [reply src/containers/Cluster/ClusterOverview/utils.tsx:42]: This was an intentional behavior change.

The goal of this PR is to align overview percentage formatting with the requirement from issue #3977: show one decimal place only for values below 1%, and show whole percentages otherwise. The current cluster and tenant overview cards now follow the same rule consistently.

File: src/utils/metrics/formatMetricLegend.ts [hunk 3]
```diff
@@ -12,11 +18,40 @@ export interface MetricFormatParams {
...
+    if (!Number.isFinite(absoluteConvertedValue) || absoluteConvertedValue < 1) {
+        return decimalPlaces;
+    }
+
+    return String(Math.trunc(absoluteConvertedValue)).length + decimalPlaces;
+}
+
 export function formatStorageLegend({value, capacity}: MetricFormatParams): string {
-    const formatted = formatStorageValues(value, capacity, undefined, '\n');
+    const size = getConsistentMetricBytesSize([value, capacity]);
```
[the remark this answers is not in the document]
Comment (kkdras) [reply src/utils/metrics/formatMetricLegend.ts:34]: The shared-unit part is not new here: this formatter has already been formatting value and capacity in the unit derived from the total value.

What changed in this PR is the unit selection at rounded boundaries. Previously a value like `999_600_000` stayed in `MB`, while now it can be promoted to `GB` by the new display-size logic. So the zero-rounding edge case comes from the new rollover behavior, not from introducing shared-unit formatting itself.

**Review by astandrik — APPROVED** [review]
